account_commission/models/account_move.py

```diff
@@ -0,0 +1,251 @@
+# Copyright 2020 Tecnativa - Manuel Calero
+# Copyright 2014-2022 Tecnativa - Pedro M. Baeza
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from lxml import etree
+
+from odoo import _, api, exceptions, fields, models
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    commission_total = fields.Float(
+        string="Commissions",
+        compute="_compute_commission_total",
+        store=True,
+    )
+    partner_agent_ids = fields.Many2many(
+        string="Agents",
+        comodel_name="res.partner",
+        compute="_compute_agents",
+        search="_search_agents",
+    )
+
+    @api.depends("partner_agent_ids", "invoice_line_ids.agent_ids.agent_id")
+    def _compute_agents(self):
+        for move in self:
+            move.partner_agent_ids = [
+                (6, 0, move.mapped("invoice_line_ids.agent_ids.agent_id").ids)
+            ]
+
+    @api.model
+    def _search_agents(self, operator, value):
+        ail_agents = self.env["account.invoice.line.agent"].search(
+            [("agent_id", operator, value)]
+        )
+        return [("id", "in", ail_agents.mapped("object_id.move_id").ids)]
+
+    @api.depends("line_ids.agent_ids.amount")
+    def _compute_commission_total(self):
+        for record in self:
+            record.commission_total = 0.0
+            for line in record.line_ids:
+                record.commission_total += sum(x.amount for x in line.agent_ids)
+
+    def action_post(self):
+        """Put settlements associated to the invoices in invoiced state."""
+        self.mapped("line_ids.settlement_id").write({"state": "invoiced"})
+        return super().action_post()
+
+    def button_draft(self):
+        """Check settled lines and put settlements associated to the invoices in
+        invoice when state is changed from cancel to draft.
+        """
+        self.mapped("line_ids.settlement_id").write({"state": "invoiced"})
+        return super().button_draft()
+
+    def button_cancel(self):
+        """Check settled lines and put settlements associated to the invoices in
+        exception.
+        """
+        if any(self.mapped("invoice_line_ids.any_settled")):
+            raise exceptions.ValidationError(
+                _("You can't cancel an invoice with settled lines"),
+            )
```

Comment:
IMO, this parts should stay in button_draft(). Any reason you think this should be in button_cancel()?